OJS.Workers.ExecutionStrategies/RunSpaAndExecuteMochaTestsExecutionStrategy.cs

```diff
@@ -0,0 +1,389 @@
+            var executor = this.CreateExecutor();
+            var checker = executionContext.Input.GetChecker();
+            return this.RunTests(codeSavePath, executor, checker, executionContext, result);
+        }
+
+        protected override IExecutionResult<TestResult> RunTests(
+            string codeSavePath,
+            IExecutor executor,
+            IChecker checker,
+            IExecutionContext<TestsInputModel> executionContext,
+            IExecutionResult<TestResult> result)
+        {
+            result.Results.AddRange(executionContext.Input.Tests
+                            .Select(test => this.RunIndividualTest(
+                                codeSavePath,
+                                executor,
+                                executionContext,
+                                test))
+                            .SelectMany(resultList => resultList));
+            return result;
+        }
+
+        private ICollection<TestResult> RunIndividualTest(
+            string codeSavePath,
+            IExecutor executor,
+            IExecutionContext<TestsInputModel> executionContext,
+            TestContext test)
+        {
+            var processExecutionResult = this.Execute(executionContext, executor, codeSavePath, test.Input);
+            return this.ExtractTestResultsFromReceivedOutput(processExecutionResult.ReceivedOutput, test.Id);
+        }
+
+        protected override IExecutor CreateExecutor()
+            => this.CreateExecutor(ProcessExecutorType.Standard);
+
+        private void ExtractSubmissionFiles<TInput>(IExecutionContext<TInput> executionContext)
+        {
+            this.ValidateAllowedFileExtension(executionContext);
+
+            var submissionFilePath = FileHelpers.BuildPath(this.WorkingDirectory, "temp");
+            File.WriteAllBytes(submissionFilePath, executionContext.FileContent);
+            using (ZipFile zip = ZipFile.Read(submissionFilePath))
+            {
+                zip.RemoveSelectedEntries("node_modules/*");
+                zip.Save();
+            }
+
+            FileHelpers.RemoveFilesFromZip(submissionFilePath, RemoveMacFolderPattern);
+            FileHelpers.UnzipFile(submissionFilePath, this.UserApplicationPath);
+
+            Directory.CreateDirectory(this.TestsPath);
+            Directory.CreateDirectory(this.NginxConfFileDirectory);
+        }
+
+        private void ValidateAllowedFileExtension<TInput>(IExecutionContext<TInput> executionContext)
+        {
+            var trimmedAllowedFileExtensions = executionContext.AllowedFileExtensions?.Trim();
+            var allowedFileExtensions = (!trimmedAllowedFileExtensions?.StartsWith(".") ?? false)
+                ? $".{trimmedAllowedFileExtensions}"
+                : trimmedAllowedFileExtensions;
+
+            if (allowedFileExtensions != Constants.ZipFileExtension)
+            {
+                throw new ArgumentException("Submission file is not a zip file!");
+            }
```

Comment:
This does not validate the submission, but the file extensions provided by the execution context (which comes from the trainers). I'm not sure we need such a validation, but as it is already implemented, maybe just change the message to be something like "This file extension is not allowed for the execution strategy, please contanct an administrator."